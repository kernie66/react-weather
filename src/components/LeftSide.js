import { List } from "reactstrap";

export default function LeftSide() {
  return (
    <List type="unstyled" className="outline-md">
      <li className="border-bottom outline-sm">Solen</li>
      <li className="pb-2 pt-2">
        <img src="img/weather/sunrise-clipart-lg.png"
          width="56px" height="32px" alt="Sunrise" />
        <span className="align-bottom ps-2">05:30</span>
      </li>
      <li>
        <img src="img/weather/sunset-clipart-lg.png"
          width="56px" height="32px" alt="Sunset" />
        <span className="align-bottom ps-2">20:30</span>
      </li>
      <li className="outline-sm border-bottom pt-5">Luftfuktighet</li>
      <li className="ps-1 pt-1">66%</li>
      <li className="outline-sm border-bottom pt-5">Vind</li>
      <li className="ps-1 pt-1">12.0 m/s</li>
      <li className="ps-1 pt-1">NNV</li>
    </List>
  );
};
