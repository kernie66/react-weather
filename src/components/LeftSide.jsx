import { useCurrentWeather } from '../hooks/weatherQueries.js';
import { Loader, Stack } from '@mantine/core';
import SunTimes from './Suntimes.jsx';
import Humidity from './Humidity.jsx';
import WindSpeed from './WindSpeed.jsx';

export default function LeftSide({ minHeight = '100%' }) {
  const { data: currentWeather } = useCurrentWeather();

  if (!currentWeather) {
    return <Loader size={32} color="indigo" />;
  }

  return (
    <Stack
      className="left-side--stack"
      mih={minHeight}
      align="flex-start"
      justify="space-between"
      gap="md"
    >
      <SunTimes />
      <Humidity />
      <WindSpeed />
    </Stack>
  );
}
