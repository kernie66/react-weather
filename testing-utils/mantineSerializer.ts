// @see https://github.com/thymikee/jest-preset-angular/issues/336#issuecomment-1536232102
// @see https://github.com/orgs/mantinedev/discussions/467
const attributesToClean: { [key: string]: RegExp[] } = {
  class: [/^m-.*$/],
  id: [/^mantine-.*$/],
  for: [/^mantine-.*$/],
  'aria-describedby': [/^mantine-.*$/],
  'aria-labelledby': [/^mantine-.*$/],
  'aria-controls': [/^mantine-.*$/],
};
const attributesToCleanKeys = Object.keys(attributesToClean);
const hasAttributesToClean = (attribute: Attr) =>
  attributesToCleanKeys.some((name) => attribute.name === name);

let lastCleanedNode: Element | null = null;

module.exports = {
  print: (val: Element, serialize: (v: Element) => string) => {
    const clone = val.cloneNode(true) as Element;

    for (const attr of Object.values(clone.attributes).filter(
      hasAttributesToClean
    )) {
      attr.value = attr.value
        .split(' ')
        .filter((attrValue: string) => {
          return !attributesToClean[attr.name].some((regex) =>
            regex.test(attrValue)
          );
        })
        .join(' ');
    }

    lastCleanedNode = clone;
    return serialize(clone);
  },
  test: (val: Element) =>
    val !== lastCleanedNode &&
    val.attributes !== undefined &&
    Object.values(val.attributes).some(hasAttributesToClean),
};
