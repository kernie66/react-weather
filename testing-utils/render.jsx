// ./test-utils/render.tsx
import { render as testingLibraryRender } from '@testing-library/react';
import { MantineProvider } from '@mantine/core';
// Import your theme object
import { createTheme } from '@mantine/core';
import {
  QueryClient,
  QueryClientProvider,
} from '@tanstack/react-query';

const theme = createTheme({});

const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
    logger: {
      log: console.log,
      warn: console.warn,
      error: () => {},
    },
  });

export const testQueryClient = createTestQueryClient();

export function render(ui) {
  console.log('testQueryClient', testQueryClient);
  return testingLibraryRender(<>{ui}</>, {
    wrapper: ({ children }) => (
      <MantineProvider theme={theme}>
        <QueryClientProvider client={testQueryClient}>
          {children}
        </QueryClientProvider>
      </MantineProvider>
    ),
  });
}
