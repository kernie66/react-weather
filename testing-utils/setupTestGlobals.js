// src/test-globals.ts
export const setup = () => {
  process.env.TZ = 'Europe/Stockholm';
};
