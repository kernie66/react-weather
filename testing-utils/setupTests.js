import { afterAll, beforeAll } from 'vitest';
import { server } from '../src/mocks/server.js';
import dayjs from 'dayjs';
import 'dayjs/locale/sv';
import localizedFormat from 'dayjs/plugin/localizedFormat.js';
import calendar from 'dayjs/plugin/calendar.js';

beforeAll(() => {
  server.listen();

  dayjs.locale('sv');
  dayjs.extend(localizedFormat);
  dayjs.extend(calendar);
});

afterEach(() => server.resetHandlers());

afterAll(() => server.close());
