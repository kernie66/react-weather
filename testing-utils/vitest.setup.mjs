// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom/vitest';

import { vi } from 'vitest';

const { getComputedStyle } = window;
window.getComputedStyle = (elt) => getComputedStyle(elt);
window.HTMLElement.prototype.scrollIntoView = () => {};

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: vi.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(),
    removeListener: vi.fn(),
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
});

class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}

window.ResizeObserver = ResizeObserver;
