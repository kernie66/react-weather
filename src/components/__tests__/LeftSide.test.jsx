import { render, screen, userEvent } from '../../../testing-utils';
import LeftSide from '../LeftSide.jsx';
import weatherDataJson from '../../../bruno/weather_alert_response.json';
import dayjs from 'dayjs';
import { testQueryClient } from '../../../testing-utils/render.jsx';
import { vi, expect } from 'vitest';

const weatherDate = dayjs.unix(weatherDataJson.current.dt);
const fakeTodayDate = new Date(weatherDate.toDate());
const fakeBeforeSunsetDate = new Date(weatherDate.hour(2).toDate());
const fakeAfterSunsetDate = new Date(weatherDate.hour(22).toDate());

describe('test LeftSide', () => {
  beforeEach(() => {
    testQueryClient.removeQueries();
    localStorage.clear();
  });

  afterAll(() => {
    vi.useRealTimers();
  });

  it('renders left side info at daytime', async () => {
    const user = userEvent.setup();
    vi.setSystemTime(fakeTodayDate);
    const snapshot = render(<LeftSide />);

    const suntimesButton = await screen.findByRole('button', {
      name: /soltider/i,
    });
    expect(suntimesButton).toBeInTheDocument();

    // Find the sunset and sunrise times
    const sunUpDown = await screen.findAllByText(
      /[0-2[0-9]:[0-5][0-9]/
    );
    // Test the correct order: sunset -> sunrise
    expect(sunUpDown.length).toBe(2);
    expect(sunUpDown[0]).toHaveTextContent('21:41');
    expect(sunUpDown[1]).toHaveTextContent('04:38');
    expect(screen.getByText(/luftfuktighet/i)).toBeInTheDocument();
    expect(screen.getByText(/vind/i)).toBeInTheDocument();
    expect(snapshot).toMatchSnapshot();

    // Click the suntimes button to open modal
    await user.click(suntimesButton);
    expect(await screen.findByRole('dialog')).toBeInTheDocument();
    expect(
      screen.getByText(/Solens tillstånd fredag 12 juli 2024/i)
    ).toBeInTheDocument();
    expect(screen.getByText(/soltillstånd/i)).toBeInTheDocument();
    expect(
      screen.getByText(/solens middagstid/i)
    ).toBeInTheDocument();
    expect(screen.getByText(/midnatt/i)).toBeInTheDocument();
  });

  it('renders left side info before sunset', async () => {
    vi.setSystemTime(fakeBeforeSunsetDate);
    render(<LeftSide />);

    expect(
      await screen.findByRole('button', { name: /soltider/i })
    ).toBeInTheDocument();

    // Find the sunset and sunrise times
    const sunUpDown = await screen.findAllByText(
      /[0-2[0-9]:[0-5][0-9]/
    );
    // Test the correct order: sunrise -> sunset
    expect(sunUpDown.length).toBe(2);
    expect(sunUpDown[0]).toHaveTextContent('04:37');
    expect(sunUpDown[1]).toHaveTextContent('21:41');
  });

  it('renders left side info after sunset', async () => {
    vi.setSystemTime(fakeAfterSunsetDate);
    render(<LeftSide />);

    expect(
      await screen.findByRole('button', { name: /soltider/i })
    ).toBeInTheDocument();

    // Find the sunset and sunrise times
    const sunUpDown = await screen.findAllByText(
      /[0-2[0-9]:[0-5][0-9]/
    );
    // Test the correct order: sunrise -> sunset
    expect(sunUpDown.length).toBe(2);
    expect(sunUpDown[0]).toHaveTextContent('04:38');
    expect(sunUpDown[1]).toHaveTextContent('21:40');
  });
});
