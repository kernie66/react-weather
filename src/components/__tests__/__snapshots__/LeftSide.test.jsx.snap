// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test LeftSide > renders left side info at daytime 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <style
        data-mantine-styles="classes"
      >
        @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
      </style>
      <div
        class="left-side--stack m_6d731127 mantine-Stack-root"
        style="--stack-gap: var(--mantine-spacing-md); --stack-align: flex-start; --stack-justify: space-between; min-height: 100%;"
      >
        <button
          class="mantine-focus-auto m_87cf2631 mantine-UnstyledButton-root"
          type="button"
        >
          <div
            class="m_6d731127 mantine-Stack-root"
            style="--stack-gap: calc(0.5rem * var(--mantine-scale)); --stack-align: stretch; --stack-justify: flex-start;"
          >
            <div
              class="m_4081bf90 mantine-Group-root"
              style="--group-gap: calc(0.5rem * var(--mantine-scale)); --group-align: center; --group-justify: space-between; --group-wrap: wrap;"
            >
              <img
                alt="sunset"
                class="m_9e117634 mantine-Image-root"
                src="file:///src/assets/images/clipart/sunset-clipart-lg.png"
              />
              <p
                class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
                style="font-weight: 500; text-align: right;"
              >
                21:41
              </p>
            </div>
            <div
              class="m_3eebeb36 mantine-Divider-root"
              data-orientation="horizontal"
              role="separator"
            />
            <div
              class="m_4081bf90 mantine-Group-root"
              style="--group-gap: calc(0.5rem * var(--mantine-scale)); --group-align: center; --group-justify: space-between; --group-wrap: wrap;"
            >
              <img
                alt="sunrise"
                class="m_9e117634 mantine-Image-root"
                src="file:///src/assets/images/clipart/sunrise-clipart-lg.png"
              />
              <p
                class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
                style="font-weight: 500; text-align: right;"
              >
                04:38
              </p>
            </div>
          </div>
          <span
            class="m_515a97f8 mantine-VisuallyHidden-root"
          >
            Soltider
          </span>
        </button>
        <div
          class="m_6d731127 mantine-Stack-root"
          style="--stack-gap: calc(0.25rem * var(--mantine-scale)); --stack-align: stretch; --stack-justify: flex-start;"
        >
          <p
            class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
          >
            Luftfuktighet
          </p>
          <div
            class="m_3eebeb36 mantine-Divider-root"
            data-orientation="horizontal"
            role="separator"
          />
          <p
            class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
          >
            74
            %
          </p>
        </div>
        <div
          class="m_6d731127 mantine-Stack-root"
          style="--stack-gap: calc(0.25rem * var(--mantine-scale)); --stack-align: stretch; --stack-justify: flex-start;"
        >
          <p
            class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
          >
            Vind
          </p>
          <div
            class="m_3eebeb36 mantine-Divider-root"
            data-orientation="horizontal"
            role="separator"
          />
          <p
            class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
          >
            2.6
             m/s
          </p>
          <p
            class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
          >
            S
          </p>
        </div>
      </div>
    </div>
    <div
      data-portal="true"
    >
      <div>
        <div
          class="m_9df02822 mantine-Modal-root"
          data-size="content"
          style="--modal-size: var(--modal-size-content); --modal-y-offset: 5dvh; --mb-z-index: 200;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <style
      data-mantine-styles="classes"
    >
      @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
    </style>
    <div
      class="left-side--stack m_6d731127 mantine-Stack-root"
      style="--stack-gap: var(--mantine-spacing-md); --stack-align: flex-start; --stack-justify: space-between; min-height: 100%;"
    >
      <button
        class="mantine-focus-auto m_87cf2631 mantine-UnstyledButton-root"
        type="button"
      >
        <div
          class="m_6d731127 mantine-Stack-root"
          style="--stack-gap: calc(0.5rem * var(--mantine-scale)); --stack-align: stretch; --stack-justify: flex-start;"
        >
          <div
            class="m_4081bf90 mantine-Group-root"
            style="--group-gap: calc(0.5rem * var(--mantine-scale)); --group-align: center; --group-justify: space-between; --group-wrap: wrap;"
          >
            <img
              alt="sunset"
              class="m_9e117634 mantine-Image-root"
              src="file:///src/assets/images/clipart/sunset-clipart-lg.png"
            />
            <p
              class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
              style="font-weight: 500; text-align: right;"
            >
              21:41
            </p>
          </div>
          <div
            class="m_3eebeb36 mantine-Divider-root"
            data-orientation="horizontal"
            role="separator"
          />
          <div
            class="m_4081bf90 mantine-Group-root"
            style="--group-gap: calc(0.5rem * var(--mantine-scale)); --group-align: center; --group-justify: space-between; --group-wrap: wrap;"
          >
            <img
              alt="sunrise"
              class="m_9e117634 mantine-Image-root"
              src="file:///src/assets/images/clipart/sunrise-clipart-lg.png"
            />
            <p
              class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
              style="font-weight: 500; text-align: right;"
            >
              04:38
            </p>
          </div>
        </div>
        <span
          class="m_515a97f8 mantine-VisuallyHidden-root"
        >
          Soltider
        </span>
      </button>
      <div
        class="m_6d731127 mantine-Stack-root"
        style="--stack-gap: calc(0.25rem * var(--mantine-scale)); --stack-align: stretch; --stack-justify: flex-start;"
      >
        <p
          class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
        >
          Luftfuktighet
        </p>
        <div
          class="m_3eebeb36 mantine-Divider-root"
          data-orientation="horizontal"
          role="separator"
        />
        <p
          class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
        >
          74
          %
        </p>
      </div>
      <div
        class="m_6d731127 mantine-Stack-root"
        style="--stack-gap: calc(0.25rem * var(--mantine-scale)); --stack-align: stretch; --stack-justify: flex-start;"
      >
        <p
          class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
        >
          Vind
        </p>
        <div
          class="m_3eebeb36 mantine-Divider-root"
          data-orientation="horizontal"
          role="separator"
        />
        <p
          class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
        >
          2.6
           m/s
        </p>
        <p
          class="mantine-focus-auto _outlineSingle_597f95 m_b6d8b162 mantine-Text-root"
        >
          S
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
